src/utils/ERC1024.ts

```diff
@@ -0,0 +1,105 @@
+): Eip1024EncryptedData => {
+  switch (version) {
+    case 'x25519-xsalsa20-poly1305': {
+      // generate ephemeral keypair
+      const ephemeralKeyPair = nacl.box.keyPair();
+
+      let publicKeyBytes;
+      // assemble encryption parameters
+      if (receiverPublicKey instanceof Uint8Array) {
+        publicKeyBytes = receiverPublicKey;
+      } else {
+        try {
+          publicKeyBytes = hexToBytes(receiverPublicKey);
+        } catch (error: any) {
+          throw new Error('Bad public key');
+        }
+      }
+
+      const messageBytes = utf8ToBytes(message);
+      const nonce = nacl.randomBytes(nacl.box.nonceLength);
+
+      // encrypt
+      const encryptedMessage = nacl.box.seal(
+        messageBytes,
+        nonce,
+        publicKeyBytes,
+        ephemeralKeyPair.secretKey,
+      );
+
+      // return encrypted data
+      return {
+        version: 'x25519-xsalsa20-poly1305',
+        nonce: bytesToBase64(nonce),
+        ephemPublicKey: bytesToBase64(ephemeralKeyPair.publicKey),
+        ciphertext: bytesToBase64(encryptedMessage),
+      } as Eip1024EncryptedData;
+    }
+    default:
+      throw new Error('Encryption type/version not supported');
+  }
+};
+
+const decrypt = (
+  encryptedData: Eip1024EncryptedData,
+  receiverPrivateKey: Hex | Uint8Array,
+): string => {
+  switch (encryptedData.version) {
+    case 'x25519-xsalsa20-poly1305': {
+      let secretKey;
+      if (receiverPrivateKey instanceof Uint8Array) {
+        secretKey = receiverPrivateKey;
+      } else {
+        try {
+          secretKey = hexToBytes(receiverPrivateKey);
+        } catch (error: any) {
+          throw new Error('Bad private key');
+        }
+      }
+
+      // assemble decryption parameters
+      const nonce = base64ToBytes(encryptedData.nonce);
+      const ciphertext = base64ToBytes(encryptedData.ciphertext);
+      const ephemPublicKey = base64ToBytes(encryptedData.ephemPublicKey);
+
+      // decrypt
+      const decryptedMessage = nacl.box.open(
+        ciphertext,
+        nonce,
+        ephemPublicKey,
+        secretKey,
+      );
+
+      // return decrypted msg data
+      if (!decryptedMessage) {
+        throw new Error('Decryption failed');
+      }
+      return bytesToUtf8(decryptedMessage);
+    }
+    default:
+      throw new Error('Encryption type/version not supported.');
+  }
+};
+
+/**
+ * An encryption and decryption utility matching ERC1024.
+ * It uses the `nacl.box` hybrid encryption scheme, along with an Ephemeral sender key to compute the shared secret.
+ * This snap only uses the `decrypt` method, but the encryption method is provided for completeness and testing.
+ *
+ * If we decide to add additional dependencies, this class along with the `nacl` utility could be replaced with `eth-sig-util`.
+ * @see https://github.com/ethereum/EIPs/pull/1098
+ * @see https://github.com/MetaMask/eth-sig-util/blob/main/src/encryption.ts
+ * @see https://nacl.cr.yp.to/box.html
```

Comment:
I think the biggest concern of copying or adding a new library is that it requires another round of vetting/auditing.

So if we do add this now, it means that the package is frozen until we get it audited.
Lets discuss this. I'm thinking we can go to the snaps team to figure out how we can get this audited.
- Potentially keep this as a separate branch, so it keeps the main branch  unblocked in case we need to to other changes.

Reply:
The alternative is to use [eth-sig-util](https://github.com/MetaMask/eth-sig-util) directly.
It will make this PR significantly smaller, but also make the snap bigger and a bit harder to update it in the extension/mobile, as we'll probably have to rebuild the lavamoat policies.

Reply:
True...

However funny thing is that we already have `@metamask/eth-sig-util` installed as a peer dependency (see yarn.lock) 😄

Reply:
And wow, since this is already used as a peer dependency... I think auditing won't be too much of a problem?
However you may need to confirm this.

Reply:
If that's the case, I'm happy to use that directly instead of reinventing the wheel :)
The only reason I did it this way was to avoid inflating the preinstalled snap bundle and lavamoat approval friction

Reply:
actually, it seems eth-sig-util is designed for node so I get warnings when using it in a snap :(